Allow for piping input, add pragmas for module code.

hughsk opened the pull request:
@chrisdickinson understand you're probably not keen on maintaining this set of modules - would be good to hear your thoughts. I'd be happy to either maintain the repo or fork the project if that's the case :)

The first change is to optionally enable piping text into glslify-stream instead of reading the file directly. A second boolean argument is added to the function such that:

``` javascript
require('glslify-stream')('path', true)
```

Will not read the file and return a through stream
for writing to.

The second is to add pragmas to specify module start/end points.

node_modules begin with:

``` glsl
#pragma glslify_file start module
```

And other external files:

``` glsl
#pragma glslify_file start
```

Both types of files end with:

``` glsl
#pragma glslify_file end
```

These two changes make it possible to add browserify-style transform streams to glslify, in that they can be piped
in a chain and ignore module code.

hughsk committed:
Allow for piping input, add pragmas for module code.

The first change is to optionally enable piping text
into glslify-stream instead of reading the file
directly. A second boolean argument is added to the
function such that:

    require('glslify-stream')(path, true)

Will not read the file and return a through stream
for writing to.

The second it to add pragmas to specify module start/end
points.

node_modules begin with:
    #pragma glslify_file start module
And other external files:
    #pragma glslify_file start
Both end with:
    #pragma glslify_file end

These two changes make it possible to add browserify-style
transform streams to glslify, in that they can be piped
in a chain and ignore module code.

chrisdickinson commented:
@hughsk great work.

As I understand it, the proposed pragma additions allow the following:

```
                                 -----> file 1 -----
                                /                   \
input.glsl -> glslify-stream ---------> file 2 --------- single file with guards
                                \                   /       |   
                                 -----> input ------        |
                                                            |
                                                         <transforms using guards>
                                                            |
                                                            |
                                                         bundle.glsl


```

A question: could we avoid the use of `#pragma glslify_file start/end <module>` by transforming the module to GLSL before the text even gets to glslify stream?

```


                              ----> file 1 ---> <transform> --
                             /                                \
input.glsl -> glslify-stream -----> file 2 ---> <transform> ------> bundle.glsl
                             \                                /
                              ----> input  ---> <transform> --

```

(I'm probably misunderstanding the purpose of the `#pragma glslify_file` directive inclusion!)

hughsk commented:
@chrisdickinson Yup, that sounds better! I ended up including the `glslify_file` directives instead to make it easier to wrap up from within browserify using [glslifyify](http://github.com/hughsk/glslifyify). Then I'd be able to write browserify transform streams which work with GLSL but also work when using glslify directly.

But had a bit more of a think about it today, and yeah I reckon just keeping the transforms in glslify makes more sense. It's simpler overall and I think there might be a better way of handling my problem if we do it that way - should be able to push that change tomorrow :)

hughsk committed:
Give glslify transform streams.

Transforms are handled much like browserify. You can pass a
`transform` option to the second argument, which takes an
array of transforms. You can pass string values and they'll
be resolved automatically.

The transform streams themselves are functions which take
a `filename` argument and return a through stream that
transforms the input.

Additionally, the `cwd` option has been added, which
determines where to resolve transform streams from. It
defaults to the target shader's directory.

glslify.resolve has been exposed to make it possible to mimic
glslify's file resolution, but it's probably worth pulling
out into its own module in the future.

hughsk commented:
@chrisdickinson Ok, so the commit I just pushed adds proper transform streams to `glslify-stream`. They're handled like browserify, and just passed as an array to a second `options` argument:

``` javascript
var stream = glslify(filename, {
    input: false // set to false to read the file directly
  , transform: [
    require('glslify-import')
  ]
})

stream.pipe(process.stdout)
```

Pushed [glslify-import](https://github.com/hughsk/glslify-import) as an example transform.

I've also exposed `locate_module` as `glslify.resolve` so glslify-import can use it, maybe it's better in its own module though? Otherwise, reckon this looks good to merge? Thanks! :)

hughsk committed:
fix small issues with pipeable streams

chrisdickinson commented:
Thanks! Awesome work. I'll add you as a maintainer + collab on this repo.

hughsk commented:
Awesome - thanks! :)
